cmd/kubeadm/app/util/patches/patches.go

```diff
@@ -0,0 +1,338 @@
+	}
+
+	// Always convert the target data to JSON.
+	patchedData, err = yaml.YAMLToJSON(patchTarget.Data)
+	if err != nil {
+		return err
+	}
+
+	// Iterate over the patchSets.
+	for _, patchSet := range pm.patchSets {
+		if patchSet.targetName != patchTarget.Name {
+			continue
+		}
+
+		// Iterate over the patches in the patchSets.
+		for _, patch := range patchSet.patches {
+			patchBytes := []byte(patch)
+
+			// Patch based on the patch type.
+			switch patchSet.patchType {
+
+			// JSON patch.
+			case types.JSONPatchType:
+				var patchObj jsonpatch.Patch
+				patchObj, err = jsonpatch.DecodePatch(patchBytes)
+				if err == nil {
+					patchedData, err = patchObj.Apply(patchedData)
+				}
+
+			// Merge patch.
+			case types.MergePatchType:
+				patchedData, err = jsonpatch.MergePatch(patchedData, patchBytes)
+
+			// Strategic merge patch.
+			case types.StrategicMergePatchType:
+				patchedData, err = strategicpatch.StrategicMergePatch(
+					patchedData,
+					patchBytes,
+					patchTarget.StrategicMergePatchObject,
+				)
+			}
+
+			if err != nil {
+				return errors.Wrapf(err, "could not apply the following patch of type %q to target %q:\n%s\n",
+					patchSet.patchType,
+					patchTarget.Name,
+					patch)
+			}
+			fmt.Fprintf(pm.output, "[patches] Applied patch of type %q to target %q\n", patchSet.patchType, patchTarget.Name)
+		}
+
+		// Update the data for this patch target.
+		patchTarget.Data = patchedData
+	}
+
+	return nil
+}
+
+// getTargetNameFromFilename accepts a file name and returns a known target name string,
+// or an error if the target name is unknown.
+func getTargetNameFromFilename(fileName string, knownTargets []string) (string, error) {
+	var targetName string
+
+	for _, target := range knownTargets {
+		if strings.HasPrefix(fileName, target) {
+			targetName = target
+			break
+		}
+	}
+
+	if len(targetName) == 0 {
+		return "", errors.Errorf("received file name %q, but target must be one of %v", knownTargets, fileName)
+	}
+	return targetName, nil
+}
+
+// getPatchTypeFromFilename accepts a file name and returns the patch type encoded in it.
+// For example, "etcd+merge.json" would return "merge". Returns an error on a unknown patch types.
+func getPatchTypeFromFilename(fileName string) (types.PatchType, error) {
+	idxDot := strings.LastIndex(fileName, ".")
+	if idxDot == -1 {
+		// Shift the dot index into the length of the string.
+		idxDot = len(fileName)
+	}
+
+	idxPlus := strings.Index(fileName, "+")
+	if idxPlus == -1 || idxPlus > idxDot {
+		// If the + character is missing or if its index is after the dot,
+		// just shift it into the dot index.
+		idxPlus = idxDot
+	} else {
+		// Increment the plus index to discard the + character itself.
+		idxPlus++
+	}
+
+	patchType := fileName[idxPlus:idxDot]
+	pt, ok := patchTypes[patchType]
+	if !ok {
+		return "", errors.Errorf("unknown patch type %q from file name %q", patchType, fileName)
+	}
+	return pt, nil
+}
```

Comment:
I am pretty sure that these two funcs can be implemented as a single one with the use of a single regex. The end result will be shorted and more readable.

Reply:
regex greatly simplifies this on our side and this was my initial implementation - single regexp that verifies the pattern. however, this makes it difficult for the user to discover which part of the pattern is invalid.

Reply:
I'm fine with both approaches, given that there is test coverage